### PR TITLE
feat: add ttl for component workflows and workflows

### PR DESCRIPTION
**Pull request by chalindukodikara**
This pull request introduces support for automatic cleanup of completed WorkflowRun and ComponentWorkflowRun resources by adding a configurable time-to-live (TTL) after completion. It also tracks the start and finish times for workflow runs, enabling precise calculation of when to delete them. These changes are reflected in both the Go API types and the generated CRD/Helm chart schemas.

**Automatic Workflow Run Cleanup**

* Added `ttlAfterCompletion` field to `WorkflowSpec`, `ComponentWorkflowSpec`, and their corresponding run specs, allowing users to specify how long completed runs should be retained before automatic deletion. The field supports flexible duration strings (e.g., "90d", "1h30m"). [[1]](diffhunk://#diff-c9b29dc98e164915dacea81b1eab4c2fb72854e8989681eb3dc0dab75439c825R62-R69) [[2]](diffhunk://#diff-d1ca92fc5375e383d6f60cb4e78ac7f2f61260526462d8e94f30f7cc71b28dcdR40-R46) [[3]](diffhunk://#diff-64c5cdf1d373caeff5516631f77c58fbc99be9879c7c69dc3cfc6cd195154f84R20-R28) [[4]](diffhunk://#diff-1108ffa43a641dd5fc5bc0300a1c5b0e81c08fa3d10ec6dcab47e8638e47b744R22-R29)
* Updated CRD and Helm chart schemas for `Workflow`, `ComponentWorkflow`, `WorkflowRun`, and `ComponentWorkflowRun` to document and validate the new `ttlAfterCompletion` field. [[1]](diffhunk://#diff-58cda2632d5689057e05f8f1e4ce4970fcbc56dd16f4ce7faca50aeca6d07788R133-R140) [[2]](diffhunk://#diff-bd79d0a0f29b7d3daa765901449f5153c7a43a54996b5acb309362bc8aeb2c7bR132-R139) [[3]](diffhunk://#diff-35af3dc50c0eeb9c8062607d27910994bbc9e76d7ec675ecc3450dabee0b954aR198-R204) [[4]](diffhunk://#diff-5c0e540aaab9ff472c124a2a9de2ae4b922699b1b88a79c626410af50ec5a822R197-R203) [[5]](diffhunk://#diff-dd5e01940b21752a47acf7722780c4548908510cfd91a292e1f6a2e7690f2f1fR44-R52) [[6]](diffhunk://#diff-003457a3942fb260e5bfcacefe7b29d9b20b339287c0bbcb942fee5a9b09eeefR43-R51) [[7]](diffhunk://#diff-e119f0287ed949c66da8c3e246f321c2511e0378c94a7ca585be9636fe009b77R60-R67) [[8]](diffhunk://#diff-9c375a4f71ccb1be767439d77fff89df63b7a5101abe9f1d9aaab7b9dc2874b8R59-R66)

**Workflow Run Lifecycle Tracking**

* Added `startedAt` and `finishedAt` fields to `WorkflowRunStatus` and `ComponentWorkflowRunStatus` to record when a run starts and finishes, supporting TTL-based deletion logic. [[1]](diffhunk://#diff-64c5cdf1d373caeff5516631f77c58fbc99be9879c7c69dc3cfc6cd195154f84R74-R82) [[2]](diffhunk://#diff-1108ffa43a641dd5fc5bc0300a1c5b0e81c08fa3d10ec6dcab47e8638e47b744R155-R163)
* Updated CRD and Helm chart schemas to include the new `startedAt` and `finishedAt` fields with appropriate descriptions and formats. [[1]](diffhunk://#diff-dd5e01940b21752a47acf7722780c4548908510cfd91a292e1f6a2e7690f2f1fR141-R146) [[2]](diffhunk://#diff-dd5e01940b21752a47acf7722780c4548908510cfd91a292e1f6a2e7690f2f1fR211-R215) [[3]](diffhunk://#diff-003457a3942fb260e5bfcacefe7b29d9b20b339287c0bbcb942fee5a9b09eeefR140-R145) [[4]](diffhunk://#diff-003457a3942fb260e5bfcacefe7b29d9b20b339287c0bbcb942fee5a9b09eeefR210-R214) [[5]](diffhunk://#diff-e119f0287ed949c66da8c3e246f321c2511e0378c94a7ca585be9636fe009b77R221-R226) [[6]](diffhunk://#diff-e119f0287ed949c66da8c3e246f321c2511e0378c94a7ca585be9636fe009b77R300-R304) [[7]](diffhunk://#diff-9c375a4f71ccb1be767439d77fff89df63b7a5101abe9f1d9aaab7b9dc2874b8R220-R225) [[8]](diffhunk://#diff-9c375a4f71ccb1be767439d77fff89df63b7a5101abe9f1d9aaab7b9dc2874b8R299-R303)

**Code Generation**

* Updated `DeepCopyInto` methods for the new fields to ensure correct copying of `startedAt` and `finishedAt` in the status structs. [[1]](diffhunk://#diff-d900babf741981553f1ae363bc2d0062aab6a95256dfa091b688df589e9a5c5dR1756-R1763) [[2]](diffhunk://#diff-d900babf741981553f1ae363bc2d0062aab6a95256dfa091b688df589e9a5c5dR4889-R4896)

### Related Issues
https://github.com/openchoreo/openchoreo/issues/1856

### Mini breakdown — changed files by top-level folder
- api/: 4 files (api/v1alpha1/{workflow_types.go, workflowrun_types.go, componentworkflow_types.go, componentworkflowrun_types.go})
- config/crd/bases/: 4 files (CRD base YAMLs for workflows/workflowruns/componentworkflows/componentworkflowruns)
- install/helm/openchoreo-control-plane/crds/: 4 files (Helm-packaged CRD equivalents)
- install/helm/openchoreo-build-plane/templates/: 4 files (workflow templates — ttlStrategy removed)
- internal/: ~12 files (notable: internal/controller/componentworkflowrun/controller.go +190/-34, internal/controller/workflowrun/controller.go +82/-1, internal/cmdutil/duration.go +122, internal/cmdutil/duration_test.go +291, internal/controller/workflowrun/controller_ttl_test.go +158)
- samples/: ~12 sample YAMLs updated (component-workflows, getting-started, workflows)
- Total observed affected files: ~40 (focus: API/CRD schema, controller reconcilers, duration parsing, tests, sample manifests)

### API / CRD surface changes (what changed; compatibility risk)
- Additions (all optional):
  - ttlAfterCompletion (string) added to:
    - WorkflowSpec, ComponentWorkflowSpec
    - WorkflowRunSpec, ComponentWorkflowRunSpec
    - CRD validation pattern: ^(\d+d)?(\s*\d+h)?(\s*\d+m)?(\s*\d+s)?$ (documents days/hours/minutes/seconds; examples provided)
    - Purpose: automatic deletion TTL for completed Run objects.
  - startedAt (*metav1.Time) and completedAt (*metav1.Time) added to:
    - WorkflowRunStatus, ComponentWorkflowRunStatus (CRD status fields with date-time format)
  - Task-level field renamed: FinishedAt → CompletedAt (structs and CRD task item properties)
- DeepCopy code updated to copy new timestamp fields.
- Compatibility risk: LOW (additive, optional). Note: ensure parser and CRD pattern remain aligned — if mismatch exists, runtime errors possible (LOW–MEDIUM operational risk).

### Tests added / updated
- New/updated unit tests:
  - internal/cmdutil/duration_test.go — comprehensive tests for ParseDuration and ValidateDuration (many formats/edge cases).
  - internal/controller/workflowrun/controller_ttl_test.go — tests for checkTTLExpiration (expired/non-expired, zero TTL, multiple formats).
  - workflowrun controller tests extended with TTL inheritance tests (appears duplicated in diff).
- Test gaps / missing coverage:
  - Integration / e2e tests that validate controller-driven deletion of WorkflowRun and ComponentWorkflowRun in a cluster.
  - Concurrency/race-condition tests for StartedAt/CompletedAt updates under concurrent reconciles.
  - Broader unit tests for new ComponentWorkflowRun helpers (handleCompletedWorkflow, syncExistingWorkflowRun, renderWorkflowResources, prepareWorkflowForRun) beyond controller-level coverage.

### Risk hotspots (short reasons)
1. Reconciliation-driven deletions — HIGH impact
   - checkTTLExpiration can delete Run resources; clock skew, RequeueAfter mis-calculation, or transient errors could cause premature deletion or missed cleanup.
2. Timestamp race conditions — MEDIUM
   - setStartedAtIfNeeded/setCompletedAtIfNeeded mutate status; concurrent reconciles may lead to lost/overwritten timestamps without conflict-handling tests.
3. Controller refactor surface area — MEDIUM
   - New helpers and reorganized flow in ComponentWorkflowRun reconciler increase regression risk; needs focused unit tests.
4. CRD regex vs parser alignment — LOW–MEDIUM
   - If CRD accepts formats ParseDuration rejects, users may pass CRD-valid but runtime-invalid TTLs.
5. Install/upgrade ordering & RBAC — LOW
   - CRD additions require CRD upgrades prior to controller rollout; ensure RBAC includes permission to delete Run resources for TTL cleanup.

### Overall assessment and recommendations
This PR implements end-to-end TTL support: API types and CRDs, controller TTL/timestamp handling for WorkflowRun and ComponentWorkflowRun, a flexible duration parser with unit tests, and sample updates. Changes are largely additive and optional, so compatibility risk is low. Primary operational risks are reconciler-driven deletions and timestamp concurrency.

Recommended actions before merge:
- Add integration/e2e tests exercising TTL expiry and actual deletion in-cluster.
- Add concurrency tests for StartedAt/CompletedAt updates.
- Verify ParseDuration accepts all CRD-allowed TTL strings (align parser and CRD pattern).
- Confirm operator Helm/upgrade steps to apply CRD changes before controller rollout and verify controller RBAC allows deletions.

### CODE REVIEW

File: internal/controller/workflowrun/controller.go [hunk 1]
```diff
@@ -612,3 +632,64 @@ func extractArgoStepOrderFromNodeName(nodeName string) int {
...
+
+	// Requeue to check again when TTL expires
+	requeueAfter := time.Until(expirationTime)
+	if requeueAfter > 0 {
+		logger.V(1).Info("Requeuing for TTL check", "requeueAfter", requeueAfter)
+		return true, ctrl.Result{RequeueAfter: requeueAfter}, nil
+	}
+
+	return false, ctrl.Result{}, nil
+}
```
[the remark this answers is not in the document]
Comment (chalindukodikara) [reply internal/controller/workflowrun/controller.go:675]: We anyway validate the field "// +kubebuilder:validation:Pattern=`^(\d+d)?(\s*\d+h)?(\s*\d+m)?(\s*\d+s)?$`". And if they provide any invalid TTL, we can avoid doing the clean up.

File: api/v1alpha1/componentworkflowrun_types.go [hunk 2]
```diff
@@ -144,6 +152,15 @@ type ComponentWorkflowRunStatus struct {
...
 	Tasks []WorkflowTask `json:"tasks,omitempty"`
+
+	// StartedAt is the timestamp when this workflow run started execution.
+	// +optional
+	StartedAt *metav1.Time `json:"startedAt,omitempty"`
+
+	// FinishedAt is the timestamp when this workflow run finished execution (succeeded or failed).
+	// This is used together with TTLAfterCompletion to determine when to delete the workflow run.
+	// +optional
+	FinishedAt *metav1.Time `json:"finishedAt,omitempty"`
```
Comment (VajiraPrabuddhaka) [inline api/v1alpha1/componentworkflowrun_types.go]: shouldn't this be CompletedAt ? as we are using the term `Complete` for ttl

Comment (chalindukodikara) [reply api/v1alpha1/componentworkflowrun_types.go]: Yes, that is better.

File: api/v1alpha1/componentworkflow_types.go [hunk 3]
```diff
@@ -37,6 +37,13 @@ type ComponentWorkflowSpec struct {
 	// Template variables are substituted with context and parameter values.
 	// +optional
 	Resources []ComponentWorkflowResource `json:"resources,omitempty"`
+
+	// TTLAfterCompletion defines the time-to-live for ComponentWorkflowRun instances after completion.
+	// it will be automatically deleted after this duration.
+	// Format: duration string supporting days, hours, minutes, seconds (e.g., "90d", "10d 1h 30m", "1h30m")
+	// +optional
+	// +kubebuilder:validation:Pattern=`^(\d+d)?(\s*\d+h)?(\s*\d+m)?(\s*\d+s)?$`
+	TTLAfterCompletion string `json:"ttlAfterCompletion,omitempty"`
```
Comment (VajiraPrabuddhaka) [inline api/v1alpha1/componentworkflow_types.go:48]: shall we check whether we can use metav1.Duration for this?

Comment (chalindukodikara) [reply api/v1alpha1/componentworkflow_types.go:48]: It doesn't support days
